// @heroku-cli/color/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const supports = require("supports-color");
const chalk_1 = require("chalk");
const ansiStyles = require("ansi-styles");
exports.CustomColors = {
    supports,
    // map gray -> dim because it's not solarized compatible
    gray: chalk_1.default.dim,
    grey: chalk_1.default.dim,
    attachment: chalk_1.default.cyan,
    addon: chalk_1.default.yellow,
    configVar: chalk_1.default.green,
    release: chalk_1.default.blue.bold,
    cmd: chalk_1.default.cyan.bold,
    app: (s) => exports.color.heroku(`⬢ ${s}`),
    heroku: (s) => {
        if (!exports.color.supports)
            return s;
        let has256 = exports.color.supports.has256 || (process.env.TERM || '').indexOf('256') !== -1;
        return has256 ? '\u001b[38;5;104m' + s + ansiStyles.reset.open : chalk_1.default.magenta(s);
    },
};
exports.color = new Proxy(chalk_1.default, {
    get: (chalk, name) => {
        if (exports.CustomColors[name])
            return exports.CustomColors[name];
        return chalk[name];
    },
});
exports.default = exports.color;
